### PR TITLE
chore: strengthen code review design pass

### PR DESCRIPTION
**Pull request by nabinchha**
## 📋 Summary
Strengthens the review-code skill so foundational design, interface, and consistency issues are evaluated explicitly before a new public contract becomes difficult to change.

## 🔗 Related Issue
N/A

## 🔄 Changes
- Require a contract inventory, invariant-ownership map, and comparison with neighboring APIs during scope analysis.
- Add concrete checks for module cohesion, private-symbol ownership, redundant entry points, generic-helper readability, builder vocabulary, precise input types, lifecycle-aware naming, exception scope, and opaque extension boundaries.
- Add public-contract test scenarios and exploratory complexity checks that prompt manual maintainability review.
- Treat hard-to-reverse flaws in new public, serialized, or extension contracts as Critical and prevent Ship it verdicts while foundational design questions remain unresolved.

## 🧪 Testing
- [x] Commit hooks pass
- [x] git diff --check passes
- [ ] make test passes (not run; instruction-only Markdown change)
- [ ] Unit tests added/updated (not applicable)
- [ ] E2E tests added/updated (not applicable)

The generic skill quick validator was also attempted. It rejects the repository-standard argument-hint and disable-model-invocation frontmatter keys; this change does not modify those keys.

## ✅ Checklist
- [x] Follows commit message conventions
- [x] Commits are signed off (DCO)
- [x] Architecture docs updated (not applicable; no product architecture change)

### CODE REVIEW

File: .agents/skills/review-code/SKILL.md [hunk 1]
```diff
@@ -111,6 +111,10 @@ Before diving into details, build a mental model:
 4. **Identify the primary goal** (feature, refactor, bugfix, etc.)
 5. **Note cross-cutting concerns** (e.g., a rename that touches many files vs. substantive logic changes)
 6. **Check existing feedback** (PR mode): inspect both inline comments (Step 1, item 5) and PR-level review bodies (Step 1, item 5b) so you don't duplicate feedback already given
+7. **Classify the contract impact**: note whether the change introduces or modifies public/exported symbols, config or model schemas, serialized formats, extension/plugin boundaries, or builder APIs. If none apply, skip items 8-10 and continue to Step 3.5.
+8. **Inventory the changed contract**: list each new or modified public symbol and public model field with its expected callers
```
Comment (andreatnvidia) [inline .agents/skills/review-code/SKILL.md]: The contract classification is broader than this inventory: serialized formats and extension boundaries may not introduce a public symbol or model field, while `public/exported` does not distinguish supported exports from unprefixed internal symbols. That can make the review miss a JSON or entry-point contract, or escalate an internal helper as a compatibility blocker. The inventory needs to cover every changed contract element and record its exposure, stability, and expected callers before severity is assigned.

File: .agents/skills/review-code/SKILL.md [hunk 2]
```diff
@@ -166,19 +173,44 @@ Re-read the changed files with a focus on **structure and design of the new/modi
...
-- Single responsibility — does each new function/class do one thing?
+- Single responsibility — does each new function, class, and module do one coherent job?
+- If one module owns distinct phases such as authored config, resolution, compilation, and validation, would splitting those responsibilities make ownership and future changes clearer?
 - Are new dependencies flowing in the right direction?
 - Could this introduce circular imports or unnecessary coupling?
+- Does every symbol live in the module that owns it? A leading-underscore helper imported by sibling modules is a misplaced shared API, not truly private.
+- Are helpers scoped to their actual consumers? Logic used only to validate or construct one class usually belongs on that class instead of in module-level generic machinery.
+- Are re-exports intentional and sourced from the defining module rather than incidental import chains?
+- Is there one canonical public entry point per operation, or do class methods plus module-level wrappers expose redundant APIs?
+- Is each invariant enforced once at the narrowest authoritative layer, rather than repeated or recomputed across models, resolution, compilation, and validation?
```
Comment (andreatnvidia) [inline .agents/skills/review-code/SKILL.md]: Could we soften `enforced once` here? An input boundary and a post-compilation boundary can both need checks for the same logical invariant because the representation or trust level changed. One authoritative owner for the business rule, with complementary boundary validation allowed, would avoid treating useful defensive checks as duplication.

File: .agents/skills/review-code/SKILL.md [hunk 3]
```diff
@@ -284,12 +327,14 @@ Call out 2-3 things done well (good abstractions, thorough tests, clean refactor
...
-Choose the verdict that matches the **highest severity** finding in the review:
+Choose the verdict that matches the **highest severity confirmed finding** in the review. If there are no confirmed Critical or Warning findings but a design question cannot be resolved without team input, use **Needs discussion**.
 
 - **Ship it** — No findings. Ready to merge as-is.
 - **Ship it (with nits)** — Only Suggestions (see above — style improvements, simplifications, or optional enhancements). Nothing blocking.
 - **Needs changes** — Any Critical or Warning findings. List the items that must be addressed before merge.
-- **Needs discussion** — Architectural or design questions that need team input before a decision can be made.
+- **Needs discussion** — No confirmed Critical or Warning findings, but an architectural or design question needs team input before a decision can be made.
+
+A confirmed foundational contract flaw is Critical and requires **Needs changes**. Reserve **Needs discussion** for uncertainty that requires team input. Do not use any **Ship it** verdict (including **Ship it (with nits)**) while either remains unresolved.
```
Comment (andreatnvidia) [inline .agents/skills/review-code/SKILL.md]: As written, any unresolved design question forbids a Ship it verdict, even when the question cannot affect merge readiness. That turns ordinary residual uncertainty into a veto despite there being no confirmed Critical or Warning finding. Could we require the question to be material to compatibility or the merge decision, and leave lower-impact uncertainty in the residual-risk section?

File: .agents/skills/review-code/SKILL.md (hunk 1 above)
Comment (nabinchha) [reply .agents/skills/review-code/SKILL.md]: Good catch. Step 3 now distinguishes supported public symbols/exports from internal names and inventories every changed contract element, including serialized representations, extension entry points, and builder operations. The inventory also records exposure, stability commitment, and expected callers or consumers. Addressed in c466719a.

File: .agents/skills/review-code/SKILL.md (hunk 3 above)
Comment (nabinchha) [reply .agents/skills/review-code/SKILL.md]: Agreed. Needs discussion is now limited to unresolved questions material to compatibility or merge readiness. I also added an optional Residual Risk section for lower-impact uncertainty and clarified that those notes do not preclude either Ship it verdict. Addressed in c466719a.

File: .agents/skills/review-code/SKILL.md (hunk 2 above)
Comment (nabinchha) [reply .agents/skills/review-code/SKILL.md]: Agreed. I reworded this to require one authoritative owner for each business invariant while explicitly allowing complementary boundary checks when the representation or trust level changes. That keeps defensive validation from being treated as duplication. Addressed in c466719a.

File: .agents/skills/review-code/SKILL.md [hunk 4]
```diff
@@ -166,19 +173,44 @@ Re-read the changed files with a focus on **structure and design of the new/modi
 
 - Does the change fit the existing architecture and patterns?
 - Are new abstractions at the right level? (too abstract / too concrete)
-- Single responsibility — does each new function/class do one thing?
+- Single responsibility — does each new function, class, and module do one coherent job?
+- If one module owns distinct phases such as authored config, resolution, compilation, and validation, would splitting those responsibilities make ownership and future changes clearer?
 - Are new dependencies flowing in the right direction?
 - Could this introduce circular imports or unnecessary coupling?
+- Does every symbol live in the module that owns it? A leading-underscore helper imported by sibling modules is a misplaced shared API, not truly private.
```
Comment (andreatnvidia) [inline .agents/skills/review-code/SKILL.md]: Could we treat a sibling import of `_helper` as a signal rather than proof that the helper is misplaced? The repository already uses package-private sharing for `interface.results._export_*`, consumed by both `results.py` and `composite_workflow.py` without exporting those helpers publicly. The categorical rule can drive unnecessary utility-module moves or accidental public exposure; checking ownership and package-export intent would distinguish the maintainability problem from a legitimate package-private boundary.

File: .agents/skills/review-code/SKILL.md [hunk 5]
```diff
@@ -282,14 +325,20 @@ Separate each finding with a blank line. Use bold file-and-title as a heading li
...
 ### Verdict
 
-Choose the verdict that matches the **highest severity** finding in the review:
+Choose the verdict that matches the **highest severity confirmed finding** in the review. If there are no confirmed Critical or Warning findings but a design question material to compatibility or merge readiness cannot be resolved without team input, use **Needs discussion**.
 
 - **Ship it** — No findings. Ready to merge as-is.
 - **Ship it (with nits)** — Only Suggestions (see above — style improvements, simplifications, or optional enhancements). Nothing blocking.
 - **Needs changes** — Any Critical or Warning findings. List the items that must be addressed before merge.
-- **Needs discussion** — Architectural or design questions that need team input before a decision can be made.
+- **Needs discussion** — No confirmed Critical or Warning findings, but an architectural or design question material to compatibility or merge readiness needs team input before a decision can be made.
```
Comment (andreatnvidia) [inline .agents/skills/review-code/SKILL.md:344]: `Needs discussion` is now reachable, but the exact template gives its material open question nowhere to live. A Warning forces `Needs changes`, a Suggestion is optional, and Residual Risk is restricted to lower-impact uncertainty, so CI can emit an unexplained verdict or force a false severity. Please authorize the question under Verdict or add an Open Questions section used with this verdict.

File: .agents/skills/review-code/SKILL.md [hunk 6]
```diff
@@ -332,7 +381,8 @@ In branch mode, skip this step — display the review to the user and note the t
 - Issues that are supposed to be caught by CI (linter, typechecker, formatter) — mention "run `make check-all`" if relevant, but don't list every style nit
 - Pre-existing issues on unmodified lines
 - Pedantic nits that don't affect correctness or maintainability
-- Intentional functionality or API changes that are clearly documented
+- Missing comments or docstrings when behavior and the public contract are otherwise clear; flag the underlying ambiguity when they are not
```
Comment (andreatnvidia) [inline .agents/skills/review-code/SKILL.md]: `STYLEGUIDE.md` requires docstrings for every public API class and function, including a one-liner for simple APIs. Ruff does not enable pydocstyle and the Pass 3 checklist omits docstrings, so this exemption can silently accept a documented project-standard violation. Could we limit it to private helpers and prose nits while keeping the public API requirement?

File: .agents/skills/review-code/SKILL.md [hunk 7]
```diff
@@ -202,6 +234,8 @@ Final pass focused on **project conventions and test quality for new/modified co
 - Are edge cases tested?
 - Are mocks/stubs used appropriately (at boundaries, not deep internals)?
 - Do new test names clearly describe what they verify?
+- For new public contracts, do tests exercise direct supported construction as well as builder/factory paths, lifecycle transitions, malformed boundary input, and an unfamiliar extension/plugin case where applicable?
+- For normalized exceptions, do tests assert the stable public error and verify that the chained cause preserves useful underlying detail?
```
Comment (andreatnvidia) [inline .agents/skills/review-code/SKILL.md]: This check assumes every normalized exception retains a chained cause, but `engine/models/errors.py` deliberately uses `raise ... from None` for many provider errors and preserves selected detail in `FormattedLLMErrorMessage`; `test_provider_error_without_cause_has_no_chain` also treats an absent chain as supported. Requiring a chain here would flag the repository's canonical pattern and can retain raw third-party exceptions that may contain sensitive request data. The check needs to accept either a safe chained cause or sanitized detail on the public error, according to the boundary's policy.

File: .agents/skills/review-code/SKILL.md (hunk 4 above)
Comment (nabinchha) [reply .agents/skills/review-code/SKILL.md]: Good point. I reworded this as a design signal rather than proof of misplacement: reviewers now inspect ownership and package-export intent, and the guidance explicitly allows legitimate unexported package-private sharing. Addressed in 0cd64eb9.

File: .agents/skills/review-code/SKILL.md (hunk 6 above)
Comment (nabinchha) [reply .agents/skills/review-code/SKILL.md]: Agreed. The skill now preserves STYLEGUIDE.md's public API docstring requirement, adds it to the Pass 3 standards checklist, and limits the exemption to clear private helpers and prose-only nits. Addressed in 0cd64eb9.

File: .agents/skills/review-code/SKILL.md (hunk 7 above)
Comment (nabinchha) [reply .agents/skills/review-code/SKILL.md]: Good catch. The boundary and test guidance now accept either a safe chained cause or sanitized detail on the public error. It explicitly allows raise ... from None when retaining a raw provider exception could expose sensitive data. Addressed in 0cd64eb9.

File: .agents/skills/review-code/SKILL.md (hunk 5 above)
Comment (nabinchha) [reply .agents/skills/review-code/SKILL.md:344]: Agreed. I added an Open Questions section used only with Needs discussion. It now requires the material question, why it affects compatibility or merge readiness, and the team decision needed. Addressed in 0cd64eb9.

**Review by andreatnvidia — APPROVED** [review]
Thanks for working through the feedback. The updates look good to me!
